### PR TITLE
added support for Day of Week in snowflake

### PR DESCRIPTION
**Pull request by rohitgaurDatametica**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 1]
```diff
@@ -229,7 +229,17 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
...
+      case "EEEE":
+      case "E4":
+      case "EEE":
+      case "E3":
+        writer.print("'DY'");
+        break;
+      default:
+        call.operand(0).unparse(writer, leftPrec, rightPrec);
+        break;
+      }
```
Comment (pratik163) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:242]: wrap it in a separate method

**Review by pratik163 — APPROVED** [review]

**Review by deepakDatametica — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 2]
```diff
@@ -288,6 +284,34 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
...
+    case "Sunday":
+    case "Monday":
+    case "Tuesday":
+    case "Wednesday":
+    case "Thursday":
+    case "Friday":
+    case "Saturday":
+      final SqlWriter.Frame operatorFrame = writer.startFunCall(call.getOperator().toString());
+      final SqlWriter.Frame substrFrame = writer.startFunCall("SUBSTR");
+      writer.print(call.operand(0).toString());
```
Comment (deepakDatametica) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:310]: replace it with `call.operand(0).unparse(writer, call, leftPrec, rightPrec)`

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 1 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 2 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:310]: done

**Review by deepakDatametica — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 3]
```diff
@@ -288,6 +284,34 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
...
+    default:
+      super.unparseCall(writer, call, leftPrec, rightPrec);
+    }
+  }
+
+  private void getDayofWeek(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    switch (((SqlLiteral) call.operand(0)).getValueAs(String.class)) {
+    case "Sunday":
+    case "Monday":
+    case "Tuesday":
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:303]: Please don't override lower function based on specific function changes

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 4]
```diff
@@ -588,6 +612,24 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
   }
 
+  private void formatDate(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    final SqlWriter.Frame formatDate = writer.startFunCall("TO_VARCHAR");
+    call.operand(1).unparse(writer, leftPrec, rightPrec);
+    writer.print(",");
+    switch (((SqlLiteral) call.operand(0)).getValueAs(String.class)) {
+    case "EEEE":
+    case "E4":
+    case "EEE":
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:622]: Please use DATE_TIME_FORMAT_MAP to convert from EEEE to DY

**Review by soma-mondal — COMMENTED** [review]
Please add test cases for format_date

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 5]
```diff
@@ -588,6 +605,53 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
     return ((SqlBasicCall) aggCall).operand(0);
   }
 
+  private void formatDate(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:608]: Please rename to unparseFormatDate

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 6]
```diff
@@ -588,6 +605,53 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
+    if (call.operand(0).toString().equals("'EEEE'") || call.operand(0).toString().equals("'E4'")) {
+      SqlCall toVarcharNode = createVarCharNode(call);
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:610]: toVarcharNode -> toVarcharCall

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 7]
```diff
@@ -588,6 +605,53 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
+      List<String> abvList = Arrays.asList("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat");
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:611]: abvList -> abvWeekDays

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 8]
```diff
@@ -588,6 +605,53 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
+      ArrayList<String> abvDays = new ArrayList<>(abvList);
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:612]: Can we iterate on List<String> abvList directly. Do we need another list here ?

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 9]
```diff
@@ -588,6 +605,53 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
+              "Wednesday", "Thursday", "Friday", "Saturday");
+      ArrayList<SqlLiteral> weekDaysNode = new ArrayList<>();
+
+      weekDays.forEach(it ->
+              weekDaysNode.add(
+                      SqlLiteral.createCharString(it, SqlParserPos.ZERO)));
+
+      SqlNodeList thenList = new SqlNodeList(weekDaysNode, SqlParserPos.ZERO);
+
+      SqlCall caseNode = new SqlCase(SqlParserPos.ZERO, null, whenList, thenList, null);
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:632]: caseNode -> caseCall

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 10]
```diff
@@ -588,6 +605,53 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
+      unparseCall(writer, caseNode, leftPrec, rightPrec);
+    } else {
+      if (DATE_TIME_FORMAT_MAP.containsKey(unquoteStringLiteral(call.operand(0).toString()))) {
+        unparseCall(writer, createVarCharNode(call), leftPrec, rightPrec);
+      } else {
+        SqlNode formatDate = SqlLibraryOperators.TO_VARCHAR.createCall(
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:638]: formatDate -> toVarcharCall

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 11]
```diff
@@ -588,6 +605,53 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
+                SqlParserPos.ZERO, call.operand(1), call.operand(0));
+
+        formatDate.unparse(writer, leftPrec, rightPrec);
+      }
+    }
+  }
+
+  private SqlCall createVarCharNode(SqlCall call) {
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:646]: Rename to createToVarcharCall

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 12]
```diff
@@ -353,7 +353,7 @@ private SqlLibraryOperators() {
 
   /** The "DAYNAME(datetime)" function; returns the name of the day of the week,
    * in the current locale, of a TIMESTAMP or DATE argument. */
-  @LibraryOperator(libraries = {MYSQL})
+  @LibraryOperator(libraries = {MYSQL, SNOWFLAKE})
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:356]: Please revert this code

**Review by soma-mondal — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 8 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:612]: we can write it like this - new ArrayList<>(Arrays.asList("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"))

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 3 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 4 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 5 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 6 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 7 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 8 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 9 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 10 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 11 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 12 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: done

**Review by soma-mondal — APPROVED** [review]

**Pull request merged**